staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldmanager
+
+import (
+	"context"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var equalities = func() conversion.Equalities {
+	// Clone equality.Semantic
+	e := conversion.EqualitiesOrDie()
+	for key, value := range equality.Semantic.Equalities {
+		// Can't call e.AddFunc on value, reflection has type-erased it and
+		// AddFunc throws error. So we directly set the value in the map
+		e.Equalities[key] = value
+	}
+
+	// Add a comparator for managed fields in addition to semantic equality
+	if err := e.AddFunc(func(a, b metav1.ManagedFieldsEntry) bool {
+		// Two objects' managed fields are equivalent if, ignoring timestamp,
+		//	the objects are deeply equal.
+		a.Time = nil
+		b.Time = nil
+
+		return reflect.DeepEqual(a, b)
+	}); err != nil {
+		panic(err)
+	}
+
+	return e
+}()
+
+// IgnoreManagedFieldsTimestampsTransformer cancels the change if the objects
+// are semantically equal (ignoring the managed fields timestamp).
+func IgnoreManagedFieldsTimestampsTransformer(
+	_ context.Context,
+	newObj runtime.Object,
+	oldObj runtime.Object,
+) (runtime.Object, error) {
+	if equalities.DeepEqual(newObj, oldObj) {
+		return oldObj, nil
```

Comment:
to my knowledge, we've never made semantic equality a condition of persisting an update before... are we sure there are no objects that make distinctions that would be invisible to this check (`null` vs `[]`, etc)

cc @sttts @deads2k

Reply:
As discussed with Jordan, we can make a strict equality with merely ignoring managed fields timestamp, which would be less intrusive.

Reply:
another clarification: these are internal objects or storage version objects? We do ignore updates of the later, we don't ignore updates of the former, or neither of request versions. It would break storage version migrator.

Reply:
Ah, that's a good point. I'm curious why no test caught that?

None-the-less, rather than cancel the write (and make a strong decision), we could merely revert the updated timestamp in the managed fields. The etcd layer can then decide if it's worth writing the object.

Reply:
I've adjusted the code to just revert the timestamp changes in managed fields, if that is all that has changed. (Rather than re-using the old object in its entirety).

Reply:
yeah, I think conceptually this is doing a lot less than before (it only changes managedFields, not the rest of the object), which is much less risky.

Reply:
To be honest, as we discussed with Alex, this was not preventing storage version migrator from updating the object since this was working on internal object, and not really by-passing the write, but keeping the same copy of the object.